ssa/manager_apply.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2021 Stefan Prodan
+Copyright 2021 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ssa
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Apply performs a server-side apply of the given object if the matching in-cluster object is different or if it doesn't exist.
+// Drift detection is performed by comparing the server-side dry-run result with the existing object.
+// When immutable field changes are detected, the object is recreated if 'force' is set to 'true'.
+func (m *ResourceManager) Apply(ctx context.Context, object *unstructured.Unstructured, force bool) (*ChangeSetEntry, error) {
+	existingObject := object.DeepCopy()
+	_ = m.client.Get(ctx, client.ObjectKeyFromObject(object), existingObject)
+
+	dryRunObject := object.DeepCopy()
+	if err := m.dryRunApply(ctx, dryRunObject); err != nil {
+		if force && strings.Contains(err.Error(), "immutable") {
```

Comment:
How safe is this check? Could I create a false positive by e.g. creating a resource with the name `immutable`, and then let it fail because of something not related to immutability?

Reply:
Yes but for now there is typed error for immutable errors, I'll add a comment here to drop the string contains once Flux min req becomes Kubernetes 1.21.

Reply:
Was not specifically looking for a typed error, but maybe an `AND` condition check which confirms another part of the common error message is present in the string after the `Contains(err, "immutable")` passes.